lib/pbio/platform/ev3/platform.c

```diff
@@ -633,6 +645,82 @@ static void Edma3CCErrHandlerIsr(void) {
+#define MMU_SECTION_SZ      (1 << MMU_SECTION_SHIFT)
+#define MMU_L1_ENTS         (1 << (32 - 20))
+#define MMU_L1_ALIGN        (16 * 1024)
+#define MMU_L1_SECTION(addr, ap, domain, c, b)      (       \
+    ((addr) & ~(MMU_SECTION_SZ - 1)) |                      \
+    (((ap) & 3) << 10) |                                    \
+    (((domain) & 0xf) << 5) |                               \
+    (1 << 4) |                                              \
+    ((!!(c)) << 3) |                                        \
+    ((!!(b)) << 2) |                                        \
+    (1 << 1)                                                \
+    )
+static uint32_t l1_page_table[MMU_L1_ENTS] __attribute__((aligned(MMU_L1_ALIGN)));
+#define SYSTEM_RAM_SZ_MB    64
+
+static void mmu_init(void) {
+    // Invalidate TLB
+    __asm__ volatile (
+        "movs r0, #0\n"
+        "mcr p15, 0, r0, c8, c7, 0"
+        ::: "r0"
+        );
+
+    // Program domain D0 = no access, D1 = manager (no permission checks)
+    // We generally don't bother with permission checks, anything valid is RWX
+    __asm__ volatile (
+        "movs r0, #0b1100\n"
+        "mcr p15, 0, r0, c3, c0, 0"
+        ::: "r0"
+        );
+
+    // For simplicity, everything is mapped as sections (1 MiB chunks)
+    // This potentially fails to catch certain out-of-bounds accesses,
+    // but as a tradeoff we do not need any L2 sections.
+
+    // MMIO register ranges
+    l1_page_table[0x01C00000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0x01C00000, 0, 1, 0, 0);
+    l1_page_table[0x01D00000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0x01D00000, 0, 1, 0, 0);
+    l1_page_table[0x01E00000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0x01E00000, 0, 1, 0, 0);
+    l1_page_table[0x01F00000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0x01F00000, 0, 1, 0, 0);
+
+    // On-chip RAM, which is used to share control structures with the PRUs
+    l1_page_table[0x80000000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0x80000000, 0, 1, 0, 0);
+
+    // Off-chip main DDR RAM
+    for (unsigned int i = 0; i < SYSTEM_RAM_SZ_MB; i++) {
+        uint32_t addr = 0xC0000000 + i * MMU_SECTION_SZ;
+        // TODO: Enable caching once DMA code is upgraded to handle cache
+        l1_page_table[addr >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(addr, 0, 1, 0, 0);
+    }
+    // Off-chip main DDR RAM, uncacheable mirror @ 0xD0000000
+    for (unsigned int i = 0; i < SYSTEM_RAM_SZ_MB; i++) {
+        uint32_t addr_phys = 0xC0000000 + i * MMU_SECTION_SZ;
+        uint32_t addr_virt = 0xD0000000 + i * MMU_SECTION_SZ;
+        l1_page_table[addr_virt >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(addr_phys, 0, 1, 0, 0);
+    }
+
+    // ARM local RAM, interrupt controller
+    l1_page_table[0xFFF00000 >> MMU_SECTION_SHIFT] = MMU_L1_SECTION(0xFFF00000, 0, 1, 0, 0);
+
+    // Make sure this all makes its way into memory
+    pbdrv_compiler_memory_barrier();
+
+    // Set TTBR
+    __asm__ volatile ("mcr p15, 0, %0, c2, c0, 0" "" : : "r" (l1_page_table));
```

Comment:
Would be nice to use `cp15.h` instead of having inline assembly here (and add functions there if needed).